ilmsdump/gui/__init__.py

```diff
@@ -0,0 +1,119 @@
+
+import ilmsdump
+
+target_path = '.'
+
+
+def addTextBrowser(text_browser, text):
+    text_browser.setText(text_browser.toPlainText() + text)
+
+
+async def get_enrolled_courses(client, log):
+    courses = []
+    try:
+        await [courses.append(i) async for i in client.get_enrolled_courses()]
+    except TypeError:
+        pass
+    for i in courses:
+        log(i.name + '\n')
+    log('\n == finish == \n\n')
+    return courses
+
+
+# TODO
+
+
+class DownloadThread(QThread):
+    def __init__(self):
+        QThread.__init__(self)
+        self.log = None
+        self.courses = None
+        self.downloader = None
+
+    def set(self, log, courses, downloader):
+        self.log = log
+        self.courses = courses
+        self.downloader = downloader
+
+    def run(self):
+        self.log(f"download is starting... {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n")
+        self.downloader.run(self.courses, [])
+        self.log(f"download finished. {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n")
+
+
+class Form(QObject):
+    def __init__(self, ui_file, parent=None):
+        # load ui file
+        super(Form, self).__init__(parent)
+        ui_file = QFile(ui_file)
+        ui_file.open(QFile.ReadOnly)
+        loader = QUiLoader()
+        self.window = loader.load(ui_file)
+        ui_file.close()
+
+        # bind widget
+        self.logBrowser = self.window.findChild(QTextBrowser, "log")
+        self.account = self.window.findChild(QLineEdit, "account")
+        self.password = self.window.findChild(QLineEdit, "password")
+        self.test_login_btn = self.window.findChild(QPushButton, "testLoginBtn")
+        self.show_enrolled_btn = self.window.findChild(QPushButton, "showEnrolledBtn")
+        self.download_btn = self.window.findChild(QPushButton, "downloadBtn")
+
+        # connect
+        self.test_login_btn.clicked.connect(self.test_login)
+        self.show_enrolled_btn.clicked.connect(self.show_enrolled)
+        self.download_btn.clicked.connect(self.download)
+
+        self.loop = asyncio.get_event_loop()
+        self.client = ilmsdump.Client(target_path)
+        self.download_thread = DownloadThread()
+        self.window.show()
+
+    def test_login(self):
+        self.client = ilmsdump.Client(target_path)
+        account = self.account.text()
+        password = self.password.text()
+        try:
+            self.loop.run_until_complete(
+                self.client.login_with_username_and_password(account, password)
+            )
+            self.log('login success\n')
+            return True
+        except ilmsdump.LoginFailed:
+            self.log('login fail\n')
+            return False
+
+    def show_enrolled(self):
+        self.log('== showing the enrolled courses == \n')
+        if not self.test_login():
+            return
+        self.courses = self.loop.run_until_complete(get_enrolled_courses(self.client, self.log))
+
+    def log(self, text):
```

Comment:
好像很多地方都會 self.log(f'... {datetime.now().strftime(...)}')
```suggestion
    def log(self, text, *, with_timestamp=False):
```

Reply:
這另外抽出成一個 function 沒問題！